ngraph_bridge/ngraph_pipelined_tensors.h

```diff
@@ -0,0 +1,175 @@
+    }
+  }
+
+  void return_index(size_t id) {
+    if (m_depth == 0) {
+      throw std::runtime_error(
+          "Depth=0, so no one should be calling return_index");
+    } else {
+      if (id > m_depth - 1) {
+        throw std::runtime_error("Depth = " + to_string(m_depth) +
+                                 " but passed an index to return ( = " +
+                                 to_string(id) + "), which is too large");
+      } else {
+        if (is_free(id)) {
+          throw std::runtime_error(
+              "Attempted to return index " + to_string(id) +
+              " but it is already present in the free indices set");
+        }
+      }
+    }
+    insert_to_free_set(id);
+  }
+
+  int get_index() {
+    if (m_depth == 0) {
+      return -1;
+    }
+    std::lock_guard<std::mutex> lock(m_mtx);
+    if (m_free_depth_indexes.size() == 0) {
+      return -1;
+    } else {
+      auto itr = m_free_depth_indexes.begin();
+      int retval = *itr;
+      m_free_depth_indexes.erase(itr);
+      return retval;
+    }
+  }
+
+  // TODO: if needed implement get_depth() and get_num_free_idxs()
+  // Implementing get_depth() might make some sense because if one receives an
+  // IndexLibrary object that only gives return_index()==-1 then one might want
+  // to know is there any point in waiting for it (it will never return anything
+  // other than -1 if depth==0). So the user of the object can query depth and
+  // throw an error or take appropriate steps if its 0
+
+ private:
+  set<int> m_free_depth_indexes;
+  size_t m_depth;
+  std::mutex m_mtx;
+
+  void insert_to_free_set(size_t id) {
+    std::lock_guard<std::mutex> lock(m_mtx);
+    m_free_depth_indexes.insert(id);
+  }
+
+  bool is_free(size_t id) {
+    std::lock_guard<std::mutex> lock(m_mtx);
+    if (id > m_depth - 1) {
+      throw std::runtime_error("Asked to check if id=" + to_string(id) +
+                               " is free, but depth=" + to_string(m_depth));
+    }
+    return m_free_depth_indexes.find(id) != m_free_depth_indexes.end();
+  }
+};
+
+class PipelinedTensorsStore {
+ public:
+  PipelinedTensorsStore(PipelinedTensorMatrix in, PipelinedTensorMatrix out)
+      : m_in_tensors(in),
+        m_out_tensors(out),
+        m_num_inputs(in.size()),
+        m_num_outputs(out.size()) {
+    if (in.size() > 0) {
+      m_depth = in[0].size();
+    } else if (out.size() > 0) {
+      m_depth = out[0].size();
+    } else {
+      // The executable has no inputs and outputs
+      m_depth = 0;
+    }
+    idx_lib = make_shared<IndexLibrary>(m_depth);
+  }
+
+  // returns a tuple of idx, and 2 vectors of ng tensors (input and output
+  // groups). If the idx is negative, then its an invalid group (because
+  // pipeline is filled right now)
+  tuple<int, PipelinedTensorVector, PipelinedTensorVector> get_tensors() {
+    int i = idx_lib->get_index();
+    return make_tuple(i, (i < 0 ? PipelinedTensorVector{} : get_group(true, i)),
```

Comment:
Instead of true/false for Input/Output tensors can you use something more descriptive, an enum like we discussed?

Reply:
Shrestha changed name of boolean `input`->`is_input`. i think that suffices.